src/iter/collect/consumer.rs

```diff
@@ -1,66 +1,97 @@
-use super::super::noop::*;
 use super::super::plumbing::*;
+use std::marker::PhantomData;
 use std::ptr;
 use std::slice;
-use std::sync::atomic::{AtomicUsize, Ordering};
 
 pub(super) struct CollectConsumer<'c, T: Send> {
-    /// Tracks how many items we successfully wrote. Used to guarantee
-    /// safety in the face of panics or buggy parallel iterators.
-    ///
-    /// In theory we could just produce this as a `CollectConsumer::Result`,
-    /// folding local counts and reducing by addition, but that requires a
-    /// certain amount of trust that the producer driving this will behave
-    /// itself. Since this count is important to the safety of marking the
-    /// memory initialized (`Vec::set_len`), we choose to keep it internal.
-    writes: &'c AtomicUsize,
-
     /// A slice covering the target memory, not yet initialized!
     target: &'c mut [T],
 }
 
 pub(super) struct CollectFolder<'c, T: Send> {
-    global_writes: &'c AtomicUsize,
-    local_writes: usize,
+    /// The folder writes into `result` and must extend the result
+    /// up to exactly this number of elements.
+    final_len: usize,
 
-    /// An iterator over the *uninitialized* target memory.
-    target: slice::IterMut<'c, T>,
+    /// The current written-to part of our slice of the target
+    result: CollectResult<'c, T>,
 }
 
+unsafe impl<'a, T> Send for CollectFolder<'a, T> where T: Send {}
+
 impl<'c, T: Send + 'c> CollectConsumer<'c, T> {
     /// The target memory is considered uninitialized, and will be
-    /// overwritten without dropping anything.
-    pub(super) fn new(writes: &'c AtomicUsize, target: &'c mut [T]) -> Self {
-        CollectConsumer { writes, target }
+    /// overwritten without reading or dropping existing values.
+    pub(super) fn new(target: &'c mut [T]) -> Self {
+        CollectConsumer { target }
+    }
+}
+
+/// CollectResult represents an initialized part of the target slice.
+///
+/// This is a proxy owner of the elements in the slice; when it drops,
+/// the elements will be dropped, unless its ownership is released before then.
+#[must_use]
+pub(super) struct CollectResult<'c, T> {
+    start: *mut T,
+    len: usize,
+    invariant_lifetime: PhantomData<&'c mut &'c mut [T]>,
+}
```

Comment:
One could wonder, since this is the initialized part, if we can't use `&mut [T]` here. But strictly following unsafe coding guidelines, we'd need a separate raw pointer to write through from the folder, since we can't use the slice's pointer to write to the element one past the end. Then we'd have to juggle two pointers, one to use for writing in the folder and one to use for representing the proxy ownership. This could maybe be ok, if the scopes of use are nesting correctly, but harder to reason about?